dts: bindings: gpio: Introduce Arduino Nano header

Marco-Peter opened the pull request:
Add Nano header connector defined by Arduino. This allows
hardware with compatible headers to define the related GPIOs.

Marco-Peter committed:
boards: stm32: gpio: Extend NUCLEO-L031K6 with Arduino Nano header

Add GPIO mapping for Arduino Nano connector on the NUCLEO-L031K6 board
from ST.

Signed-off-by: Marco Peter <marco.peter@joylab.ch>

erwango commented on boards/arm/arduino_nano_33_iot/arduino_nano_r3_connector.dtsi:
```diff
@@ -0,0 +1,39 @@
+	};
+};
+
+arduino_i2c: &sercom4 {};
+arduino_spi: &sercom1 {};
+arduino_serial: &sercom5 {};
```
I think this becomes `arduino_nano_foo`

Marco-Peter replied:
Done

Marco-Peter committed:
boards: stm32: gpio: Extend Nucleo-G031K8 with Arduino Nano header

Add GPIO mapping for Arduino Nano connector on the NUCLEO-G031K8 board
from ST.

Signed-off-by: Marco Peter <marco.peter@joylab.ch>

mbolivar-nordic commented on dts/bindings/gpio/arduino-nano-header-r3.yaml:
```diff
@@ -0,0 +1,38 @@
+      signals labeled from A0 through A7, as well a digital signal D13. The
+      power supply and reset pins are not exposed by this binding.
+
+    This binding provides a nexus mapping for 22 pins where parent pins 0
+    through 7 correspond to A0 through A7, and parent pins 8 through 21
+    correspond to D0 through D13, as depicted below.
```
Can you confirm that this number is consistent with the pin numbering used by pinMode() (https://www.arduino.cc/reference/en/language/functions/digital-io/pinmode/) for these boards?

Marco-Peter replied:
The pin numbering on the Nucleos is correct.
Doing a crosscheck using the Arduino Studio showed that on the Arduino Nano, D0 and D1 were interchanged. On the board itself those are marked as "RX0" and "TX1".
pinMode() takes as argument "0".."13" for the digital pins and "A0".."A7" for the analog pins.

mbolivar-nordic replied:
> pinMode() takes as argument "0".."13" for the digital pins and "A0".."A7" for the analog pins.

Yes, but A0 through A7 are just macros for pin numbers. I am asking if the numeric values you are assigning to pins here in this GPIO nexus node match the same numeric values that pinMode() expects.

Marco-Peter replied:
Adapted assignments, since nobody seemed to object.

Marco-Peter committed:
boards: arduino: gpio: Extend Nano 33 IoT with Arduino Nano header

Add GPIO mapping for Arduino Nano connector on the Nano 33 IoT board
from Arduino.

Signed-off-by: Marco Peter <marco.peter@joylab.ch>

MaureenHelm reviewed (changes requested):
Please squash the fixup changes into the original commits

Marco-Peter committed:
dts: bindings: gpio: Introduce Arduino Nano header

Add Nano header connector defined by Arduino. This allows
hardware with compatible headers to define the related GPIOs.

Signed-off-by: Marco Peter <marco.peter@joylab.ch>

erwango commented:
@Marco-peter would you mind rebase and see if fixing CI issue ?

Marco-Peter commented:
Compilation of samples/drivers/flash_shell still fails on arduino_nano_33_iot:

Here the relevant output:
```
INFO    - 500/[577](https://github.com/zephyrproject-rtos/zephyr/runs/5630973355?check_suite_focus=true#step:11:577) arduino_nano_33_iot       samples/drivers/flash_shell/sample.drivers.flash.shell FAILED Build failure (build)
INFO    - /__w/zephyr/zephyr/twister-out/arduino_nano_33_iot/samples/drivers/flash_shell/sample.drivers.flash.shell/build.log
ERROR   - Loading Zephyr default modules (Zephyr base).
```
... and then ...
```
[135/166] Building C object zephyr/drivers/flash/CMakeFiles/drivers__flash.dir/flash_sam0.c.obj
FAILED: zephyr/drivers/flash/CMakeFiles/drivers__flash.dir/flash_sam0.c.obj 
ccache /opt/toolchains/zephyr-sdk-0.13.2/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc -DKERNEL -D_FORTIFY_SOURCE=2 -D__PROGRAM_START -D__ZEPHYR_SUPERVISOR__ -D__ZEPHYR__=1 -I../../../../../../include -Izephyr/include/generated -I../../../../../../soc/arm/atmel_sam0/samd21 -I../../../../../../lib/util/fnmatch/. -I../../../../../../drivers -I../../../../../../soc/arm/atmel_sam0/common/. -I/__w/zephyr/modules/hal/cmsis/CMSIS/Core/Include -I/__w/zephyr/modules/hal/atmel/asf/sam0/include/samd21 -isystem ../../../../../../lib/libc/minimal/include -isystem /opt/toolchains/zephyr-sdk-0.13.2/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/include -isystem /opt/toolchains/zephyr-sdk-0.13.2/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/include-fixed -Os -imacros /__w/zephyr/zephyr/twister-out/arduino_nano_33_iot/samples/drivers/flash_shell/sample.drivers.flash.shell/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -gdwarf-4 -fdiagnostics-color=always -mcpu=cortex-m0plus -mthumb -mabi=aapcs -mfp16-format=ieee -imacros /__w/zephyr/zephyr/include/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -Werror -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-reorder-functions -fno-defer-pop -fmacro-prefix-map=/__w/zephyr/zephyr/samples/drivers/flash_shell=CMAKE_SOURCE_DIR -fmacro-prefix-map=/__w/zephyr/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/__w/zephyr=WEST_TOPDIR -ffunction-sections -fdata-sections -std=c99 -nostdinc -MD -MT zephyr/drivers/flash/CMakeFiles/drivers__flash.dir/flash_sam0.c.obj -MF zephyr/drivers/flash/CMakeFiles/drivers__flash.dir/flash_sam0.c.obj.d -o zephyr/drivers/flash/CMakeFiles/drivers__flash.dir/flash_sam0.c.obj -c /__w/zephyr/zephyr/drivers/flash/flash_sam0.c
In file included from ../../../../../../soc/arm/atmel_sam0/samd21/soc.h:34,
                 from /__w/zephyr/zephyr/drivers/flash/flash_sam0.c:17:
/__w/zephyr/zephyr/drivers/flash/flash_sam0.c: In function 'flash_sam0_write_protection':
/__w/zephyr/modules/hal/atmel/asf/sam0/include/samd21/samd21g18a.h:535:31: error: implicit declaration of function '_UL' [-Werror=implicit-function-declaration]
  535 | #define FLASH_SIZE            _UL(0x00040000) /* 256 kB */
      |                               ^~~
/__w/zephyr/zephyr/drivers/flash/flash_sam0.c:36:27: note: in expansion of macro 'FLASH_SIZE'
   36 | #define LOCK_REGION_SIZE (FLASH_SIZE / LOCK_REGIONS)
      |                           ^~~~~~~~~~
/__w/zephyr/zephyr/drivers/flash/flash_sam0.c:394:17: note: in expansion of macro 'LOCK_REGION_SIZE'
  394 |       offset += LOCK_REGION_SIZE) {
      |                 ^~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
```
The macro`_UL` which is used by the HAL module in `/zephyrproject/modules/hal/atmel/asf/sam0/include/samd21/samd21g18a.h` seems not to be defined anywhere accessible...

Should I create an issue for this?

mbolivar-nordic commented:
> The macro`_UL` which is used by the HAL module in `/zephyrproject/modules/hal/atmel/asf/sam0/include/samd21/samd21g18a.h` seems not to be defined anywhere accessible...

@nandojve any advice here?

nandojve commented:
> > The macro`_UL` which is used by the HAL module in `/zephyrproject/modules/hal/atmel/asf/sam0/include/samd21/samd21g18a.h` seems not to be defined anywhere accessible...
> 
> @nandojve any advice here?

Hi @mbolivar-nordic ,

There is a fix for this at hal_atmel/19, which is part of Atmel pinctrl series #43731. If possible, I would like to have pinctrl merged first.

mbolivar-nordic commented:
> If possible, I would like to have pinctrl merged first.

I certainly won't object to that. I gave some review on the pinctrl PR

nandojve commented:
Hi @Marco-Peter ,

The #43731 got merged. Could you rebase?

jfischer-no commented on dts/bindings/gpio/arduino-nano-header-r3.yaml:
```diff
@@ -0,0 +1,38 @@
+
+    The Arduino Nano layout provides two headers on opposite edges of the board.
+
+    * A 15-pin header with mostly digital signals. The additional NRST (pin3)
```
In case you need to push one more time, please fix `The additional NRST (pin 3)`

Marco-Peter commented:
@nandojve, @mbolivar-nordic 
There still seems to be a problem when building `samples/drivers/flash_shell`:
```
In file included from /workdir/zephyr/boards/arm/arduino_nano_33_iot/arduino_nano_33_iot.dts:9,
                 from <command-line>:
/workdir/zephyr/boards/arm/arduino_nano_33_iot/arduino_nano_33_iot-pinctrl.dtsi:6:10: fatal error: dt-bindings/pinctrl/samd21-da1gXabcd-pinctrl.h: No such file or directory
    6 | #include <dt-bindings/pinctrl/samd21-da1gXabcd-pinctrl.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```
The file `dt-bindings/pinctrl/samd21-da1gXabcd-pinctrl.h` is referenced from a few other files too, so I assume it should have been added with https://github.com/zephyrproject-rtos/zephyr/pull/43731. Is this correct?

nandojve commented:
> @nandojve, @mbolivar-nordic There still seems to be a problem when building `samples/drivers/flash_shell`:
> 
> ```
> In file included from /workdir/zephyr/boards/arm/arduino_nano_33_iot/arduino_nano_33_iot.dts:9,
>                  from <command-line>:
> /workdir/zephyr/boards/arm/arduino_nano_33_iot/arduino_nano_33_iot-pinctrl.dtsi:6:10: fatal error: dt-bindings/pinctrl/samd21-da1gXabcd-pinctrl.h: No such file or directory
>     6 | #include <dt-bindings/pinctrl/samd21-da1gXabcd-pinctrl.h>
>       |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> compilation terminated.
> ```
> 
> The file `dt-bindings/pinctrl/samd21-da1gXabcd-pinctrl.h` is referenced from a few other files too, so I assume it should have been added with #43731. Is this correct?

Yes, correctly! Did you update modules locally using `west update` ?

Marco-Peter commented:
> > @nandojve, @mbolivar-nordic There still seems to be a problem when building `samples/drivers/flash_shell`:
> > ```
> > In file included from /workdir/zephyr/boards/arm/arduino_nano_33_iot/arduino_nano_33_iot.dts:9,
> >                  from <command-line>:
> > /workdir/zephyr/boards/arm/arduino_nano_33_iot/arduino_nano_33_iot-pinctrl.dtsi:6:10: fatal error: dt-bindings/pinctrl/samd21-da1gXabcd-pinctrl.h: No such file or directory
> >     6 | #include <dt-bindings/pinctrl/samd21-da1gXabcd-pinctrl.h>
> >       |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> > compilation terminated.
> > ```

> >   
> > The file `dt-bindings/pinctrl/samd21-da1gXabcd-pinctrl.h` is referenced from a few other files too, so I assume it should have been added with #43731. Is this correct?
> 
> Yes, correctly! Did you update modules locally using `west update` ?

This did it, thanks

mbolivar-nordic commented:
CI seems stuck here. @stephanosio any idea?

stephanosio commented:
> CI seems stuck here. @stephanosio any idea?

we had yet another GitHub mishap today.
